### DISCUSSION TITLE
Encoding atomic properties in SMILES

### DISCUSSION THREAD

**Discussion by AlanKerstjens** [rdkit/rdkit#4185] [Q&A] [answered]
Dear RDKit community,

I'm storing molecules in a SQL database and trying to enforce their uniqueness. However, I'd like to not only compare their molecular graphs but also an unsigned integer property of some atoms. 

Up until now I was doing this by encoding the property as the mass number of a connected pseudoatom and comparing the canonical SMILES. For example, if methane's carbon had property value 18 I'd have the following SMILES: `[18*]C`

The problem with this approach is that a `RDKit::Atom` stores its mass number as a `std::uint16_t`, and I'd like to encode a `std::uint64_t` instead.

What would be the most elegant way of doing this? 

If there were a way of mapping atom indices to SMILES characters I could perform the string edition myself. As far as I know there is no RDKit function for this. In a different discussion Greg suggested doing it with regex (https://github.com/rdkit/rdkit/discussions/4151).

InChIs are the other solution that springs to mind. Can one encode arbitrary properties in InChI strings/keys?

Any suggestions would be much appreciated!

**Comment by bp-kelley** [comment]
If you want to use smiles, atom maps are stored and parsed as uint32, but this isn’t portable.

see

atom.Set/GetAtomMapNum

otherwise you may have to store string properties on the atom as we don’t expose int64 ( you could save two int32s though)

In python

atom.SetProp(“prop”, repr(v))

 pickle of the molecule.  Note you will have to turn atom properties on for pickling first.

Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.AllProps)

**Comment by bp-kelley** [comment]
By portable, I mean between different smiles parsers.

**Comment by greglandrum** [comment, marked as answer]
As @bp-kelley mentioned: the only "extra" properties which are taken into account when canonicalizing the molecular graph are the atom map numbers and those are stored as int32 (not uint32).

So if you want the property to be used during the canonicalization and can't fit it into an int32 then there's nothing in the RDKit which can help you.

If you are ok with the normal canonicalization code (i.e the properties don't determine the output atom order) and want to just store the properties together with the SMILES, then you can assign atom properties and use CXSMILES. The properties don't affect the canonicalization, but they are output in canonical order.
```
In [6]: m = Chem.MolFromSmiles('COCCC')

In [7]: for atom in m.GetAtoms():
   ...:     atom.SetProp("p1",str(0xf00df00d+atom.GetIdx()+1))
   ...: 

In [8]: Chem.MolToCXSmiles(m)
Out[8]: 'CCCOC |atomProp:0.p1.4027445266:1.p1.4027445265:2.p1.4027445264:3.p1.4027445263:4.p1.4027445262|'
```
Note that the CXSmiles generation code is currently pretty verbose and will include every atomic  and molecular property in the output, so you may need to do a bit of cleanup of properties before calling `MolToCXSmiles()` in your case.

**Reply by AlanKerstjens** [reply]
Thanks @greglandrum. I think CXSMILES will be a good solution for my case since I don't need the properties to influence the canonicalization of the molecular graph.
